util/cpuprofile/cpuprofile.go

```diff
@@ -133,12 +135,15 @@ func (p *parallelCPUProfiler) register(ch ProfileConsumer) {
 		return
 	}
 	p.Lock()
+	isFirst := len(p.cs) == 0
 	p.cs[ch] = struct{}{}
 	p.Unlock()
 
-	select {
-	case p.notifyRegister <- struct{}{}:
-	default:
+	if isFirst {
```

Comment:
There is a race condition there that the isFirst in L138 is not the same as the actual "is first" when you use it at L142.

Reply:
Sorry, I can't get you, there is already a lock, why still have race condition?

Reply:
Suppose:

- When executed at L138 there is one registered downstream, so that your `isFirst = false`. 
- During executing L140 ~ L142, this downstream is deregistered.
- Then at L143 `notifyRegister` is not called as you have `isFirst = false` but at that time there is already no downstream any more.

Reply:
Ok, look like the `isFirst` variable is misleading, I already remove it, and use another way to implement this logic, PTAL